File: 2707. Extra Characters in a String
/Solution.py
class Solution:
    def minExtraChar(self, s: str, dictionary: List[str]) -> int:
        dp = [0 for i in range(51)]
        for i in range(len(s) - 1, -1, -1):
            dp[i] = 1 + dp[i + 1]
            for d in dictionary:
                if s[i:i + len(d)] == d:
                    dp[i] = min(dp[i], dp[i + len(d)])
        return dp[0]

File: 2707. Extra Characters in a String
/Solution.java
class Solution {
    int[] dp = new int[51];
    public int minExtraChar(String s, String[] dictionary) {
        return recurse(s, dictionary, 0);
    }
    public int recurse(String s, String[] dict, int i) {
        if (i == s.length()) return 0;
        if (dp[i] != 0) return dp[i] - 1;
        dp[i] = 2 + recurse(s, dict, i + 1);
        for (String d : dict) {
            if (i + d.length() <= s.length() && s.substring(i, i + d.length()).equals(d)) {
                dp[i] = Math.min(dp[i], 1 + recurse(s, dict, i + d.length()));
            }
        }
        return dp[i] - 1;
    }
}
